lib/lambda.js

```diff
@@ -2,11 +2,311 @@
+  const message = record && record.Sns
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.type = 'pubsub'
+  faas.trigger.request_id = record && record.Sns && record.Sns.MessageId
+
+  trans.setFaas(faas)
+
+  let topicArn = record && record.Sns && record.Sns.TopicArn
+  topicArn = topicArn || ''
+  const arnParts = topicArn.split(':')
+  const topicName = arnParts[5]
+  const accountId = arnParts[4]
+  const region = arnParts[3]
+
+  trans.name = `RECEIVE ${topicName}`
+  trans.type = 'messaging'
+
+  const serviceContext = {
+    origin: {
+      name: topicName,
+      id: topicArn,
+      version: record.EventVersion
+    }
+  }
+  trans.setServiceContext(serviceContext)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws',
+      region: region,
+      service: {
+        name: 'sns'
+      },
+      account: {
+        id: accountId
+      }
+    }
+  }
+  trans.setCloudContext(cloudContext)
+
+  let age
+  if (message.Timestamp) {
+    age = Date.now() - (new Date(message.Timestamp).getTime())
+  }
+  const messageContext = {
+    queue: {
+      name: topicName
+    },
+    age: {
+      ms: age
+    },
+    body: undefined,
+    headers: undefined
+  }
+  if (agent._conf.captureHeaders) {
+    messageContext.headers = Object.assign({}, message.MessageAttributes)
+  }
+
+  if (shouldCaptureBody(agent)) {
+    messageContext.body = message.Message
+  }
+  trans.setMessageContext(messageContext)
+}
+
+function isSnsEvent (event) {
+  const records = event.Records ? event.Records : []
+  return records.length === 1 && records[0].EventSource === 'aws:sns'
+}
+
+function setS3SingleData (trans, event, context, isColdStart) {
+  const record = event.Records[0]
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.type = 'datasource'
+  faas.trigger.request_id = record.responseElements && record.responseElements['x-amz-request-id']
+
+  trans.setFaas(faas)
+  trans.name = `${record && record.eventName} ${record && record.s3 && record.s3.bucket && record.s3.bucket.name}`
+  trans.type = 'request'
+
+  const serviceContext = {
+    origin: {
+      name: record && record.s3 && record.s3.bucket && record.s3.bucket.name,
+      id: record && record.s3 && record.s3.bucket && record.s3.bucket.arn,
+      version: record.eventVersion
+    }
+  }
+  trans.setServiceContext(serviceContext)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws',
+      service: {
+        name: 's3'
+      },
+      region: record.awsRegion
+    }
+  }
+  trans.setCloudContext(cloudContext)
+}
+
+function isS3SingleEvent (event) {
+  const records = event.Records ? event.Records : []
+  return records.length === 1 && records[0].eventSource === 'aws:s3'
+}
+
+function setTransactionResultFromApiGatewayResponse (trans, result, event) {
+  if (!result) {
+    return
+  }
+  const statusCode = result.statusCode ? result.statusCode : result.statuscode
+  if (statusCode >= 500 && isApiGatewayEvent(event)) {
```

Comment:
I asked in slack if the Python impl isn't guarding on `statusCode >= 500` here. However, that can come afterwards if different -- and with clarification of the spec.